Bottom: dbfe4089fc3fd1267152d2c8034771cbb9f90712
Top:    6bba2754f862a3b8a8031db5c7b74c60047fff84
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-04-17 07:46:51 +0800

tuntap: tweak on the path of non-xdp case in tun_build_skb()

If we're sure not to go native XDP, there's no need for several things
like bh and rcu stuffs.

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index 1e44a4d..380ab64 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1674,10 +1674,12 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 	 * of xdp_prog above, this should be rare and for simplicity
 	 * we do XDP on skb in case the headroom is not enough.
 	 */
-	if (hdr->gso_type || !xdp_prog)
+	if (hdr->gso_type || !xdp_prog) {
 		*skb_xdp = 1;
-	else
-		*skb_xdp = 0;
+		goto build;
+	}
+
+	*skb_xdp = 0;
 
 	local_bh_disable();
 	rcu_read_lock();
@@ -1723,6 +1725,7 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 	rcu_read_unlock();
 	local_bh_enable();
 
+build:
 	skb = build_skb(buf, buflen);
 	if (!skb) {
 		skb = ERR_PTR(-ENOMEM);
